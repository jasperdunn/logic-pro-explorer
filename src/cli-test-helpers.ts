import { execSync, spawnSync } from 'child_process'
import { dirname, resolve } from 'path'
import { fileURLToPath } from 'url'

export const packagePath = resolve(dirname(fileURLToPath(import.meta.url)), '../dist/index.js')

export function runCommand(command: string): string {
  return execSync(`node ${packagePath} ${command}`).toString()
}

export function runSpawnedCommand(command: string, input = ''): string {
  const result = spawnSync('node', [packagePath, command], { input })
  if (result.status !== 0) {
    throw new Error(`Command '${command}' failed with exit code ${result.status}.`)
  }
  return result.stdout.toString()
}
