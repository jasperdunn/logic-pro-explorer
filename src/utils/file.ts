import fs from 'fs/promises'
import { resolve, join } from 'path'
import type { PathLike } from 'fs'
import { isFinderAlias } from './mac.js'
import { ErrorCode, errorMessage, getErrorMessage, isErrorCode } from './error.js'

export const packageProperties = await getPackageProperties()

export async function getFilePaths(
  extension: FileExtension,
  directoryPath: string,
  limit?: number
): Promise<string[]> {
  const filePaths: string[] = []

  try {
    if (await isFinderAlias(directoryPath)) {
      // should this be an exception?
      throw new Error(`This path is an alias. Please use the resolved directory path.`)
    }

    if (!(await fs.stat(directoryPath)).isDirectory()) {
      // should this be an exception?
      throw new Error('The path is not a directory.')
    }

    const stack = [directoryPath]

    while (stack.length > 0 && (!limit || filePaths.length < limit)) {
      const currentPath = stack.pop()

      if (!currentPath) {
        continue
      }

      const pathParts = await fs.readdir(currentPath)

      for (const part of pathParts.reverse()) {
        const fullPath = join(currentPath, part)

        if (limit && filePaths.length >= limit) {
          break
        }

        if (part.endsWith(extension)) {
          filePaths.push(fullPath)
        }

        if ((await fs.stat(fullPath)).isDirectory()) {
          stack.push(fullPath)
        }
      }
    }
  } catch (error) {
    if (isErrorCode(error, ErrorCode.ENOENT)) {
      throw new Error(errorMessage.directoryDoesNotExist(directoryPath))
    }

    throw new Error(
      `An error occurred while reading the directory: "${directoryPath}"\n${getErrorMessage(error)}`
    )
  }

  return filePaths
}

export function getFileExtensionFromType(type: FileType): FileExtension {
  switch (type) {
    case 'project':
      return '.logicx'

    case 'component':
      return '.component'
  }
}

type PackageProperties = {
  name: string
  version: string
  description: string
}
/**
 * Required until node officially supports importing JSON files.
 */
async function getPackageProperties(): Promise<PackageProperties> {
  try {
    return JSON.parse(
      await fs.readFile(resolve('package.json'), { encoding: 'utf-8' })
    ) as PackageProperties
  } catch (error) {
    throw new Error(
      `An error occurred while reading the package.json file.\n${getErrorMessage(error)}`
    )
  }
}

export async function doesPathExist(url: PathLike): Promise<boolean> {
  try {
    await fs.access(url)
    return true
  } catch (error) {
    if (isErrorCode(error, ErrorCode.ENOENT)) {
      return false
    }

    console.error(error)

    throw error
  }
}

export const fileTypes = ['project', 'component'] as const
export type FileType = (typeof fileTypes)[number]
export type FileExtension = '.logicx' | '.component'
