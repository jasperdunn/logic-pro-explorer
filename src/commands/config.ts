import os from 'os'
import fs from 'fs/promises'
import path from 'path'
import { createCommand } from '@commander-js/extra-typings'
import { getErrorMessage } from '../utils/error.js'
import { iterativeMerge } from '../utils/object.js'
import type { DeepPartial } from '../types.js'
import type { FileType } from '../utils/file.js'
import { doesPathExist, packageProperties } from '../utils/file.js'

const { name } = packageProperties

export const projectDirectoryFlags = '-p, --project-directory <path>'
export const componentDirectoryFlags = '-c, --component-directory <path>'

export const configPath = `/Users/${os.userInfo().username}/.${name}/config.json`
const defaultProjectDirectory = `/Users/${os.userInfo().username}/Music/Logic/`
const defaultComponentDirectory = '/Library/Audio/Plug-Ins/Components/'

export const config = await getConfig()

export const configCommand = createCommand('config')
  .description('Set configuration options.')
  .option(projectDirectoryFlags, 'Set the directory to scan for Logic projects.')
  .option(componentDirectoryFlags, 'Set the directory to scan for components.')
  .option('-r, --reset', 'Reset the configuration to the default values.')
  .action(async ({ projectDirectory, componentDirectory, reset }) => {
    if (reset) {
      await initializeConfig({ reset: true })
      return
    }

    if (projectDirectory || componentDirectory) {
      await updateConfig({
        directory: {
          component: componentDirectory,
          project: projectDirectory,
        },
      })
    }
  })

type Config = {
  directory: Record<FileType, string>
}

type InitializeConfigOptions = {
  reset: true
}
async function initializeConfig(options?: InitializeConfigOptions): Promise<void> {
  try {
    if (!options?.reset && (await doesPathExist(configPath))) {
      return
    }

    const configDirectoryPath = path.dirname(configPath)
    if (!(await doesPathExist(configDirectoryPath))) {
      await fs.mkdir(configDirectoryPath)
    }

    const configData: Config = {
      directory: {
        project: defaultProjectDirectory,
        component: defaultComponentDirectory,
      },
    }

    await fs.writeFile(configPath, JSON.stringify(configData, null, 2))
  } catch (error) {
    throw new Error(
      `An error occurred while initializing the config file.\n${getErrorMessage(error)}`
    )
  }
}

async function getConfig(): Promise<Config> {
  try {
    await initializeConfig()

    return JSON.parse(await fs.readFile(configPath, { encoding: 'utf-8' })) as Config
  } catch (error) {
    throw new Error(`An error occurred while reading the config file.\n${getErrorMessage(error)}`)
  }
}

export async function updateConfig(partialConfig: DeepPartial<Config>): Promise<void> {
  try {
    await fs.writeFile(configPath, JSON.stringify(iterativeMerge(config, partialConfig), null, 2), {
      encoding: 'utf-8',
    })
  } catch (error) {
    throw new Error(`An error occurred while updating the config file.\n${getErrorMessage(error)}`)
  }
}

export async function removeConfig(): Promise<void> {
  try {
    await fs.rm(path.dirname(configPath), {
      force: true,
      recursive: true,
    })
  } catch (error) {
    throw new Error(`An error occurred while removing the config file.\n${getErrorMessage(error)}`)
  }
}
