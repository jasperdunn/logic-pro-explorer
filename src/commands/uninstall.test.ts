import { runSpawnedCommand } from '../cli-test-helpers.js'
import { doesPathExist } from '../utils/file.js'
import { configPath, config, updateConfig } from './config.js'

describe('uninstall command', () => {
  const originalConfig = structuredClone(config)
  afterAll(async () => {
    await updateConfig(originalConfig)
  })

  // TODO: figure out how to test this, the config file is being deleted whilst other tests are running
  test.skip('should prompt the user for confirmation', async () => {
    const actual = runSpawnedCommand('uninstall', 'y')

    const configFileExists = await doesPathExist(configPath)

    // why is the snapshot file generating multiple times?
    expect(actual).toMatchSnapshot()
    expect(configFileExists).toBe(false)
  })

  test('should close the app if the user declines', () => {
    const actual = runSpawnedCommand('uninstall', 'n')

    expect(actual).toMatchSnapshot()
  })
})
