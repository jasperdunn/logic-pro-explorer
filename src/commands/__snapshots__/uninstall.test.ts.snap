// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uninstall command > should close the app if the user declines 1`] = `
"? Are you sure you want to uninstall logic-pro-explorer? (Y/n) [63D[63C[2K[G? Are you sure you want to uninstall logic-pro-explorer? (Y/n) n[64D[64C[2K[G? Are you sure you want to uninstall logic-pro-explorer? No[59D[59C
"
`;
